src/io/flutter/logging/tree/DataPanel.java

```diff
@@ -0,0 +1,282 @@
+  }
+
+  class NodeModel extends BaseTreeModel<Object> {
+    private final DiagnosticsNode node;
+    private final String rootContent;
+
+    public NodeModel(DiagnosticsNode node, String rootContent) {
+      this.node = node;
+      this.rootContent = rootContent;
+    }
+
+    @Override
+    public List<? extends Object> getChildren(Object parent) {
+      if (parent instanceof DiagnosticsNode) {
+        return DataPanel.getChildren((DiagnosticsNode)parent);
+      }
+      if (parent == rootContent) {
+        return DataPanel.getChildren(node);
+      }
+      return null;
+    }
+
+    @Override
+    public Object getRoot() {
+      return rootContent;
+    }
+  }
+
+  private final Gson gsonHelper = new GsonBuilder().setPrettyPrinting().create();
+  private final HTMLEditorKit editorKit;
+
+  private final EventDispatcher<ContentListener>
+    dispatcher = EventDispatcher.create(ContentListener.class);
+
+  private DataPanel() {
+    setBorder(JBUI.Borders.empty());
+    editorKit = new HTMLEditorKit();
+  }
+
+  public static DataPanel create() {
+    final DataPanel panel = new DataPanel();
+    panel.setLayout(new BoxLayout(panel, BoxLayout.Y_AXIS));
+    return panel;
+  }
+
+  public void onUpdate(@NotNull ContentListener listener) {
+    dispatcher.addListener(listener);
+  }
+
+  private void linkSelected(URL url) {
+    // TODO(pq): handle links
+  }
+
+  public void update(Object data) {
+    // Remove stale content.
+    for (Component c : getComponents()) {
+      remove(c);
+    }
+
+    boolean showDataPane = false;
+    if (data instanceof DiagnosticsNode) {
+      showDataPane = updateNodeData((DiagnosticsNode)data);
+    }
+    else if (data instanceof String && JsonUtils.hasJsonData((String)data)) {
+      showDataPane = updateJsonTextData((String)data);
+    }
+
+    showPane(showDataPane);
+  }
+
+  private void showPane(boolean show) {
+    setVisible(show);
+    if (show) {
+      repaint();
+      revalidate();
+    }
+    dispatcher.getMulticaster().hasContent(show);
+
+    // TODO(pq): set caret position
+    // Scroll to start.
+    // setCaretPosition(0);
+  }
+
+  private boolean updateJsonTextData(String data) {
+    @SuppressWarnings("ConstantConditions") final JsonElement jsonElement = new JsonParser().parse((String)data);
+    final String text = gsonHelper.toJson(jsonElement);
+    if (text.isEmpty()) {
+      return false;
+    }
+
+    final JEditorPane editorPane = createEditorPane();
+    editorPane.setText(html(pre(text)));
+    add(editorPane);
+
+    return true;
+  }
+
+  private boolean updateNodeData(DiagnosticsNode data) {
+    // Header.
+    addText(data.toString(), data);
+
+    // Body.
+    final List<DiagnosticsNode> properties = data.getInlineProperties();
+    for (DiagnosticsNode node : properties) {
+      String contents = "";
+      if (node.getName() != null) {
+        contents += node.getName();
+      }
+
+      final ArrayList<DiagnosticsNode> children = getChildren(node);
```

Comment:
This code is missing a number of edge cases. For example, you shouldn't assume that all "name" and "description" fields should be separated by a colon. It would be best to refactor the code that renders a DiagnosticsNode as a list of icons and styled text blocks. That can work fine with your clever CSS scheme. Add a TODO and we can work on the refactor later.